llvm/lib/Transforms/IPO/SampleProfile.cpp

```diff
@@ -2460,63 +2528,108 @@ void SampleProfileMatcher::runOnFunction(const Function &F) {
+                              PreMatchStats.NumMismatchedCallsites);
+    ProfStatsVec.emplace_back("TotalProfiledCallsites",
+                              PreMatchStats.TotalProfiledCallsites);
+    ProfStatsVec.emplace_back("MismatchedCallsiteSamples",
+                              PreMatchStats.MismatchedCallsiteSamples);
+    ProfStatsVec.emplace_back("TotalProfiledFunc",
+                              PreMatchStats.TotalProfiledFunc);
+    ProfStatsVec.emplace_back("TotalFunctionSamples",
+                              PreMatchStats.TotalFunctionSamples);
     if (FunctionSamples::ProfileIsProbeBased) {
-      ProfStatsVec.emplace_back("NumMismatchedFuncHash", NumMismatchedFuncHash);
-      ProfStatsVec.emplace_back("TotalProfiledFunc", TotalProfiledFunc);
+      ProfStatsVec.emplace_back("NumMismatchedFuncHash",
+                                PreMatchStats.NumMismatchedFuncHash);
       ProfStatsVec.emplace_back("MismatchedFuncHashSamples",
-                                MismatchedFuncHashSamples);
-      ProfStatsVec.emplace_back("TotalFuncHashSamples", TotalFuncHashSamples);
+                                PreMatchStats.MismatchedFuncHashSamples);
+    }
+    if (SalvageStaleProfile) {
+      ProfStatsVec.emplace_back("PostMatchNumMismatchedCallsites",
+                                PostMatchStats.NumMismatchedCallsites);
+      ProfStatsVec.emplace_back("NumCallsitesForMatching",
+                                PostMatchStats.TotalProfiledCallsites);
+      ProfStatsVec.emplace_back("PostMatchMismatchedCallsiteSamples",
+                                PostMatchStats.MismatchedCallsiteSamples);
     }
-
-    ProfStatsVec.emplace_back("NumMismatchedCallsites", NumMismatchedCallsites);
-    ProfStatsVec.emplace_back("TotalProfiledCallsites", TotalProfiledCallsites);
-    ProfStatsVec.emplace_back("MismatchedCallsiteSamples",
-                              MismatchedCallsiteSamples);
-    ProfStatsVec.emplace_back("TotalCallsiteSamples", TotalCallsiteSamples);
 
     auto *MD = MDB.createLLVMStats(ProfStatsVec);
     auto *NMD = M.getOrInsertNamedMetadata("llvm.stats");
     NMD->addOperand(MD);
   }
 }
 
+void SampleProfileMatcher::runOnModule() {
+  ProfileConverter::flattenProfile(Reader.getProfiles(), FlattenedProfiles,
+                                   FunctionSamples::ProfileIsCS);
+  for (auto &F : M) {
+    if (ShouldSkipProfileLoading(F))
+      continue;
+    runOnFunction(F);
+  }
+
+  if (SalvageStaleProfile)
+    distributeIRToProfileLocationMap();
+
+  PreMatchStats.countMismatchedCallsiteSamples();
+  if (SalvageStaleProfile) {
+    // If a function doesn't run the matching but has mismatched callsites, this
+    // won't be any data for that function in post-match stats, so just reuse
+    // the pre-match stats.
+    PostMatchStats.copyUnchangedCallsiteMismatches(
+        PreMatchStats.FuncMismatchedCallsites);
+    PostMatchStats.countMismatchedCallsiteSamples();
```

Comment:
I see two problems: 
1. We could end up just copying the same mismatched call sites and creating duplicates.
2. The *counting* is somewhat spread out and tangled with matching -- there is `countProfileMismatches` and `countMismatchedCallsites` in function level `runStaleProfileMatching`; then there is counting again here at module level. 

How about we make these changes:
1. Move FuncMismatchedCallsites out of stat class
2. Make FuncMismatchedCallsites `StringMap<std::unordered_map<LineLocation,bool>>` with the boolean indicating whether the input mismatch was recovered by fuzzy matching. 
3. During function level `runStaleProfileMatching`, we only populate and update `FuncMismatchedCallsites`, but not doing any stats. We do stats all at the end on module level.

This is try to create clearer separation between matching and stats, and also be consistent as to at which level stats are done (function or module).

Reply:
Try to understand this on high-level, I feel this is to compress as much as we can into a global structure(`FuncMismatchedCallistes`) during function-level matching.
 Comparing to my first version(using a global IR/Prof anchor map), this use a more compact structure. Comparing to my second(this) version, even move all the `TotalProfiledFunction/ total Function Samples / total callsites` stats out of the matching.
 
That sounds good idea, just some minor thing I can think:

1) For `StringMap<std::unordered_map<LineLocation,bool>>`, a `bool` might not be enough, the matching could work badly and introduce more mismatching. may need another bool to indicate it's a new mismatch.
2) We need to save `TotalProfiledCallsite`, as we don't have the ProfAnchors for the counting at the end.

Reply:
> For StringMap<std::unordered_map<LineLocation,bool>>, a bool might not be enough, the matching could work badly and introduce more mismatching. may need another bool to indicate it's a new mismatch.

Could you explain this? what is a new mismatch? My understanding is there is a set of mismatches from input, after fuzzy matching, we recovered some of the originally mismatched, though we don't know if the recovery is correct or not.

> We need to save TotalProfiledCallsite, as we don't have the ProfAnchors for the counting at the end. 

Incrementing counter along with tracking `FuncMismatchedCallsites` seems fine?

Reply:
> Could you explain this? what is a new mismatch? My understanding is there is a set of mismatches from input, after fuzzy matching, we recovered some of the originally mismatched, though we don't know if the recovery is correct or not.

Yeah,  It depends on which heuristic we use. Some anchors are aligned based on the previous matched anchors, there could be the original function is by luck matched but we run matching to "fix" it to a mismatched location. for example:

Before matching, supposing 5:E is matched by luck.
```
IR :       1:A, 2: B,   3:C,   4: D,   5:E 
                           |       |       |  
Profile :                 3:A,   4: B,    5:E,  6:D,  7:E
```
 After matching:
 ```
IR :          1:A,  2: B,   3:C,     4: D,    5:E 
               |        |       |       |        |
Profile :    3:A,   4: B,   5:E,    6:D,     8:E
```
However, after matching, the 5:E is synced to 3:c based on the offset (1:A --> 3:A), this is the new mismatch. So, that's why I was thinking to recompute/reuse the "counting" function from scratch, without relying on the matching heuristic.

> 
> > We need to save TotalProfiledCallsite, as we don't have the ProfAnchors for the counting at the end.
> 
> Incrementing counter along with tracking `FuncMismatchedCallsites` seems fine?

OK, that seems not obey the rule to move  "all" the stats out of function-level. No strong option on this.

Reply:
I see.. Now I understand what new mismatch means -- thanks for the clarification.

There are two scenarios I think:

1. The original match by luck is actually incorrect too. In this case there is no new mismatch, but we are under-reporting input mismatches as there is no easy way to tell whether it's correct match or not. I seems the example you have is in this category. I tend to think we just need to live with it? 

2. The original match by luck is correct. In this case should we try to make sure the matching algorithm doesn't introduce new mismatches. Or at least make sure such case is very very unlikely to happen? Does this actually happen and would Meyers diffing have the same issue of introducing new mismatches? 

Regardless if we really want consider such cases, maybe `StringMap<std::unordered_map<LineLocation, MismatchState>>` with `MismatchState` containing one bit for input mismatch, and one bit for final mismatch is all we need? 

If we really want to shoot for best/most accurate metric, lucky but incorrect original match (like 1) should be counted as mismatch. But it's probably not worth the complexity to do that? 

> OK, that seems not obey the rule to move "all" the stats out of function-level. No strong option on this.

I guess strictly speaking you are right. But it still moves most of the non-trivial counting out of matching logic.

Reply:
Pushed a new version. The changes are:

Most of the counting part are moved out of the function-level matching, and moved into a new function `computeAndReportProfileStaleness`, which is a module-level function. It includes all the counts about function checksum mismatch and callsite mismatch, and both weighted(samples) and unweighted mismatches. 

Extended the (the old FuncMismatchedCallsites) to `  StringMap<std::unordered_map<LineLocation, MatchState, LineLocationHash>>
      FuncCallsiteMatchStates;`, use a enum MatchState to present the match state.  one reason is now we can compress the struture as we discussed before. Another reason is we have to move `total profiled function metrics` or `total profiled callsites` out into `computeAndReportProfileStaleness`.  it's not easy /clean to compute the two metrics in the old way because it's only under non-imported function, however, for match states, we need to count the imported function too.  So to achieve this, we need a "Matched" callsite state used to count the original profiled function.